File: src/main/java/ir/reservs/reservs/data/AppDataManager.java
package ir.reservs.reservs.data;

import java.util.List;

import io.reactivex.Single;
import ir.reservs.reservs.data.network.ApiHelper;
import ir.reservs.reservs.data.prefs.PreferencesHelper;
import ir.reservs.reservs.model.ChangePassword;
import ir.reservs.reservs.model.User;
import ir.reservs.reservs.model.ReserveHistory;
import ir.reservs.reservs.model.Salon;
import ir.reservs.reservs.model.Success;
import ir.reservs.reservs.model.Time;

public class AppDataManager implements DataManager {
    private final PreferencesHelper mPreferencesHelper;
    private final ApiHelper mApiHelper;

    public AppDataManager(PreferencesHelper preferencesHelper, ApiHelper apiHelper) {
        this.mPreferencesHelper = preferencesHelper;
        this.mApiHelper = apiHelper;
    }


    @Override
    public Single<User> register(String name, String phone, String password) {
        return mApiHelper.register(name, phone, password);
    }

    @Override
    public Single<User> login(String phone, String password) {
        return mApiHelper.login(phone, password);
    }

    @Override
    public Single<List<ReserveHistory>> reserves() {
        return mApiHelper.reserves();
    }

    @Override
    public Single<Success> reserve(String time_id, Integer salon_id, String authority, String date) {
        return mApiHelper.reserve(time_id, salon_id, authority, date);
    }

    @Override
    public Single<Success> reserveUpdate(String authority) {
        return mApiHelper.reserveUpdate(authority);
    }


    @Override
    public Single<List<Salon>> salons() {
        return mApiHelper.salons();
    }

    @Override
    public Long getCurrentUserId() {
        return mPreferencesHelper.getCurrentUserId();
    }

    @Override
    public void setCurrentUserId(Long userId) {
        mPreferencesHelper.setCurrentUserId(userId);
    }

    @Override
    public String getCurrentUserPhone() {
        return mPreferencesHelper.getCurrentUserPhone();
    }

    @Override
    public void setCurrentUserPhone(String phone) {
        mPreferencesHelper.setCurrentUserPhone(phone);
    }

    @Override
    public String getCurrentUserName() {
        return mPreferencesHelper.getCurrentUserName();
    }

    @Override
    public void setCurrentUserName(String name) {
        mPreferencesHelper.setCurrentUserName(name);
    }

    @Override
    public void setCurrentUserImage(String image) {
        mPreferencesHelper.setCurrentUserImage(image);
    }

    @Override
    public String getCurrentUserImage() {
        return mPreferencesHelper.getCurrentUserImage();
    }

    @Override
    public String getAccessToken() {
        return mPreferencesHelper.getAccessToken();
    }

    @Override
    public void setAccessToken(String token) {
        mPreferencesHelper.setAccessToken(token);
    }

    @Override
    public void removeAccessToken() {
        mPreferencesHelper.removeAccessToken();
    }

    @Override
    public Single<Success> updateName(String name) {
        return mApiHelper.updateName(name);
    }

    @Override
    public Single<ChangePassword> updatePassword(String current_password, String new_password) {
        return mApiHelper.updatePassword(current_password, new_password);
    }

    @Override
    public Single<List<Time>> times(int salon_id, String date) {
        return mApiHelper.times(salon_id, date);
    }
}

File: src/main/java/ir/reservs/reservs/data/AppDataManager.kt
package ir.reservs.reservs.data

import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import ir.reservs.reservs.data.network.ApiHelper
import ir.reservs.reservs.data.prefs.PreferencesHelper
import ir.reservs.reservs.model.*
import okhttp3.MultipartBody

class AppDataManager(private val mPreferencesHelper: PreferencesHelper, private val mApiHelper: ApiHelper) : DataManager {
    override fun confirmCode(phone: String, code: String): Single<User> {
        return config(mApiHelper.confirmCode(phone, code))
    }

    override fun sendPhone(phone: String): Single<Message> {
        return config(mApiHelper.sendPhone(phone))
    }

    override fun forget(phone: String): Single<Success> {
        return config(mApiHelper.forget(phone))
    }

    override fun updateAvatar(avatar: MultipartBody.Part): Single<Avatar> {
        return config(mApiHelper.updateAvatar(avatar))
    }

    override fun storeDevice(uniqueId: String?, push_token: String?, app_version: String?, api_level: String?,
                             operator: String?, brand: String?, model: String?): Single<Message> {
        return config(mApiHelper.storeDevice(uniqueId, push_token, app_version,
                api_level, operator, brand, model))
    }

    override fun updateFcmToken(fcmToken: String): Single<Success> {
        return config(mApiHelper.updateFcmToken(fcmToken))
    }

    override fun register(name: String, phone: String, password: String, fcmToken: String): Single<User> {
        return config(mApiHelper.register(name, phone, password, fcmToken))
    }

    override fun login(phone: String, password: String): Single<User> {
        return config(mApiHelper.login(phone, password))
    }

    override fun reserves(page: Int): Single<MutableList<History>> {
        return config(mApiHelper.reserves(page))
    }


    override fun reserve(time_id: String, salon_id: Int?, callBackUrl: String, date: String, credit: Boolean): Single<Payment> {
        return config(mApiHelper.reserve(time_id, salon_id, callBackUrl, date, credit))
    }

    override fun reserveUpdate(authority: String): Single<Success> {
        return config(mApiHelper.reserveUpdate(authority))
    }


    override fun salons(): Single<MutableList<Salon>> {
        return config(mApiHelper.salons())
    }

    override fun salons(cityId: Int): Single<MutableList<Salon>> {
        return config(mApiHelper.salons(cityId))
    }

    override fun getCurrentUserId(): Long? {
        return mPreferencesHelper.getCurrentUserId()
    }

    override fun setCurrentUserId(userId: Long?) {
        mPreferencesHelper.setCurrentUserId(userId)
    }

    override fun getCurrentUserPhone(): String {
        return mPreferencesHelper.getCurrentUserPhone()
    }

    override fun setCurrentUserPhone(phone: String?) {
        mPreferencesHelper.setCurrentUserPhone(phone)
    }

    override fun getCurrentUserName(): String {
        return mPreferencesHelper.getCurrentUserName()
    }

    override fun setCurrentUserName(name: String?) {
        mPreferencesHelper.setCurrentUserName(name)
    }

    override fun setCurrentUserAge(age: String?) {
        mPreferencesHelper.setCurrentUserAge(age)
    }

    override fun setCurrentUserImage(image: String?) {
        mPreferencesHelper.setCurrentUserImage(image)
    }

    override fun getCurrentUserImage(): String {
        return mPreferencesHelper.getCurrentUserImage()
    }

    override fun getCurrentUserAge(): String {
        return mPreferencesHelper.getCurrentUserAge()
    }

    override fun getAccessToken(): String? {
        return mPreferencesHelper.getAccessToken()
    }

    override fun setAccessToken(token: String?) {
        mPreferencesHelper.setAccessToken(token)
    }

    override fun logout() {
        mPreferencesHelper.logout()
    }

    override fun setIsVerify(status: Boolean) {
        mPreferencesHelper.setIsVerify(status)
    }

    override fun getVerify(): Boolean {
        return mPreferencesHelper.getVerify()
    }

    override fun setUserBirthday(birthday: String) {
        mPreferencesHelper.setUserBirthday(birthday)
    }

    override fun getUserBirthday(): String? {
        return mPreferencesHelper.getUserBirthday()
    }

    override fun setCredit(amount: Long) {
        mPreferencesHelper.setCredit(amount)
    }

    override fun getCredit(): Long {
        return mPreferencesHelper.getCredit()
    }

    override fun updateInformation(name: String, birthday: String?): Single<Success> {
        return config(mApiHelper.updateInformation(name, birthday))
    }

    override fun getUserInformation(): Single<User> {
        return config(mApiHelper.getUserInformation())
    }

    override fun updatePassword(current_password: String, new_password: String): Single<ChangePassword> {
        return config(mApiHelper.updatePassword(current_password, new_password))
    }

    override fun times(salon_id: Int, date: String): Single<MutableList<Time>> {
        return config(mApiHelper.times(salon_id, date))
    }

    override fun cities(): Single<MutableList<City>> {
        return config(mApiHelper.cities())
    }

    private fun <T : Any?> config(single: Single<T>): Single<T> {
        return single.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }
}
